### PR TITLE
Net: context cleanup

### PR DESCRIPTION
**Pull request by tbursztyka**
I am not going to manipulate net_context's array into a sys_list. It goes through the array only at one place: to get a free net_context. It's enough as it is, and avoids to add sys_node_t attribute into struct net_context.

### CODE REVIEW

File: include/net/net_offload.h [hunk 1]
```diff
@@ -420,6 +422,88 @@ static inline int net_offload_put(struct net_if *iface,
 	return net_if_offload(iface)->put(context);
 }
 
+#else
+static inline int net_offload_get(struct net_if *iface,
```
Comment (jukkar) [inline include/net/net_offload.h]: This cannot be done like this. The inline version must be placed together with the documentation block so that documentation is generated properly.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -119,114 +120,103 @@ int net_context_get(sa_family_t family,
...
-			if (family != AF_PACKET && family != AF_CAN) {
-				return -EINVAL;
-			}
-		} else if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET) &&
-			   !IS_ENABLED(CONFIG_NET_SOCKETS_CAN)) {
-			if (family != AF_PACKET) {
-				return -EINVAL;
+	if (IS_ENABLED(CONFIG_NET_CONTEXT_CHECK)) {
+		if (!IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+			NET_WARN("IPv4 disabled");
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: Why did you remove the asserts and used logging instead? The point in the original code was that the error cases are exceptions and user will not trigger these 99% of the time, so the warning is too much in that case.

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -1164,22 +1121,22 @@ static int context_setup_udp_packet(struct net_context *context,
...
-	    net_context_get_family(context) == AF_INET) {
-		struct sockaddr_in *addr4 = (struct sockaddr_in *)dst_addr;
-
-		dst_port = addr4->sin_port;
-
-		ret = net_context_create_ipv4_new(context, pkt,
-						  NULL, &addr4->sin_addr);
-	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
-		   net_context_get_family(context) == AF_INET6) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) &&
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: This looks like unnecessary change as it just swaps ipv4 and ipv6 blocks.

File: subsys/net/ip/net_context.c [hunk 4]
```diff
@@ -1216,12 +1173,12 @@ static void context_finalize_packet(struct net_context *context,
 
 	net_pkt_cursor_init(pkt);
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) &&
-	    net_context_get_family(context) == AF_INET) {
-		net_ipv4_finalize(pkt, net_context_get_ip_proto(context));
-	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
-		   net_context_get_family(context) == AF_INET6) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) &&
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: ditto

File: subsys/net/ip/net_context.c [hunk 5]
```diff
@@ -1289,26 +1246,26 @@ static int context_sendto(struct net_context *context,
 		return -EDESTADDRREQ;
 	}
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) &&
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: ditto

File: subsys/net/ip/net_context.c [hunk 6]
```diff
@@ -1493,12 +1450,12 @@ int net_context_send(struct net_context *context,
 		goto unlock;
 	}
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) &&
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: and here too

**Review by jukkar — CHANGES_REQUESTED** [review]

File: include/net/net_offload.h (hunk 1 above)
Comment (tbursztyka) [reply include/net/net_offload.h]: let's try with @cond INTERNAL_HIDDEN

File: subsys/net/ip/net_context.c (hunk 2 above)
Comment (tbursztyka) [reply subsys/net/ip/net_context.c]: sure it can be NET_DBG

The pb I had with the assert are:

- first the assert itself is a bit heavy as these are non-critical error (unlike a pointer that must be non null for instance. Let's say: the net_pkt pointer all through the recv path).
- second was readability: we do a test, and right below the same test, but reversed, is done for the sake of the assert. In some place it made quite a lot of lines just for the assert, generating a big amount of "useless" lines since we already did a proper test (it could have been a 'false' always put as first parameter to the net assert though)

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (tbursztyka) [reply subsys/net/ip/net_context.c]: read the commit message. 75% of the context.c have this order: af_inet6 then af_inet, and while I made the net_pkt/net_context api changed, I reversed that order in some place. So I decided to move it back.

File: subsys/net/ip/net_context.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/ip/net_context.c]: Lets use NET_DBG() then

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (jukkar) [reply subsys/net/ip/net_context.c]: There was just too many changes here that were not related -> more difficult to review. I have no issue if we reverse the if's, just put it to separate commit.

File: include/net/net_offload.h (hunk 1 above)
Comment (jukkar) [reply include/net/net_offload.h]: Then we do not get any documentation at all for these functions. We need to follow the documentation pattern like this
```
/**
 * @brief Free/close a network context.
 *
 * @details This releases the context. It is not possible to
 * send or receive data via this context after this call.
 * This is similar as BSD shutdown() function.
 *
 * @param iface Network interface where the offloaded IP stack can be
 * reached.
 * @param context The context to be closed.
 *
 * @return 0 if ok, < 0 if error
 */
static inline int net_offload_put(struct net_if *iface,
				  struct net_context *context)
{
#if defined(CONFIG_NET_OFFLOAD)
	NET_ASSERT(iface);
	NET_ASSERT(net_if_offload(iface));
	NET_ASSERT(net_if_offload(iface)->put);
	return net_if_offload(iface)->put(context);
#else 
	return 0;
#endif 
}
```
This is very ugly looking but needed because of limitations in doxygen. I just do not want to fix this one later when we generate documentation for 2.0.

File: include/net/net_ip.h [hunk 7]
```diff
@@ -278,12 +278,8 @@ struct sockaddr_storage {
 struct net_addr {
```
Comment (jukkar) [inline include/net/net_ip.h]: Typo in commit subject s/uninon/union/

I agree the network code looks a bit ugly because of ifdef's that are needed in the code. But if we try to limit the ram usage, then what you are proposing is no good. Why would we relax our own rules here, I think we should leave the ifdef's here.

Comment (tbursztyka) [reply include/net/net_ip.h]: Myeah I am also not completely convinced on that move, though as mentioned in the commit it's a really a little bloat.

**Review by jukkar — APPROVED** [review]
LGTM
